plugins/out_calyptia/calyptia.c

```diff
@@ -830,29 +862,79 @@ static void debug_payload(struct flb_calyptia *ctx, void *data, size_t bytes)
     cmt_destroy(cmt);
 }
 
-static void cb_calyptia_flush(struct flb_event_chunk *event_chunk,
-                              struct flb_output_flush *out_flush,
-                              struct flb_input_instance *i_ins,
-                              void *out_context,
-                              struct flb_config *config)
+static int cb_calyptia_exit(void *data, struct flb_config *config)
 {
-    int ret = FLB_RETRY;
-    size_t off = 0;
-    size_t out_size = 0;
-    char *out_buf = NULL;
+    struct flb_calyptia *ctx = data;
+
+    if (!ctx) {
+        return 0;
+    }
+
+    if (ctx->u) {
+        flb_upstream_destroy(ctx->u);
+    }
+
+    if (ctx->agent_id) {
+        flb_sds_destroy(ctx->agent_id);
+    }
+
+    if (ctx->agent_token) {
+        flb_sds_destroy(ctx->agent_token);
+    }
+
+    if (ctx->env) {
+        flb_env_destroy(ctx->env);
+    }
+
+    if (ctx->metrics_endpoint) {
+        flb_sds_destroy(ctx->metrics_endpoint);
+    }
 
-/* used to create records for reporting traces to the cloud. */
 #ifdef FLB_HAVE_CHUNK_TRACE
-    flb_sds_t json;
+    if (ctx->trace_endpoint) {
+        flb_sds_destroy(ctx->trace_endpoint);
+    }
 #endif /* FLB_HAVE_CHUNK_TRACE */
 
+    if (ctx->fs) {
+        flb_fstore_destroy(ctx->fs);
+    }
+
+    flb_kv_release(&ctx->kv_labels);
+    flb_free(ctx);
+
+    return 0;
+}
+
+static void cb_calyptia_flush(struct flb_event_chunk *event_chunk,
+                             struct flb_output_flush *out_flush,
+                             struct flb_input_instance *i_ins,
+                             void *out_context,
+                             struct flb_config *config)
+{
+    int ret;
+    size_t off = 0;
+    size_t out_size = 0;
+    char *out_buf = NULL;
     struct flb_connection *u_conn;
     struct flb_http_client *c = NULL;
     struct flb_calyptia *ctx = out_context;
     struct cmt *cmt;
+    flb_sds_t json;
     (void) i_ins;
     (void) config;
 
+    if (!ctx->agent_id && ctx->register_retry_on_flush) {
+        flb_plg_info(ctx->ins, "agent_id not found and register_retry_on_flush=true, attempting registration");
+        if (register_agent(ctx, config) != FLB_OK) {
+            FLB_OUTPUT_RETURN(FLB_RETRY);
+        }
+    }
+    else if (!ctx->agent_id) {
+        flb_plg_error(ctx->ins, "no agent_id available and register_retry_on_flush=false");
+        FLB_OUTPUT_RETURN(FLB_ERROR);
+    }
+
```

Comment:
Probably just a single check for agent_id then nest the retry check with different output that way? We're just checking the same thing twice otherwise.